crates/bevy_dynamic_plugin/src/loader.rs

```diff
@@ -20,15 +20,46 @@ pub enum DynamicPluginLoadError {
 /// The specified plugin must be linked against the exact same libbevy.so as this program.
 /// In addition the `_bevy_create_plugin` symbol must not be manually created, but instead created
 /// by deriving `DynamicPlugin` on a unit struct implementing [`Plugin`].
-pub unsafe fn dynamically_load_plugin(
-    path: &str,
-) -> Result<(Library, Box<dyn Plugin>), DynamicPluginLoadError> {
+pub unsafe fn dynamically_load_plugin(path: &str) -> Result<DynamicPlugin, DynamicPluginLoadError> {
     let lib = Library::new(path).map_err(DynamicPluginLoadError::Library)?;
     let func: Symbol<CreatePlugin> = lib
         .get(b"_bevy_create_plugin")
         .map_err(DynamicPluginLoadError::Plugin)?;
     let plugin = Box::from_raw(func());
-    Ok((lib, plugin))
+    Ok(DynamicPlugin(plugin, lib))
+}
+
+/// Wraps a dynamically loaded plugin and its associated library so that they can be dropped correctly.
+///
+/// It is impossible to leave dangling pointers to values and/or functions owned by a dynamic plugin.
+/// For this to hold true, the following guarantee must be met:
+/// - [`DynamicPlugin`] directly stores a [`Box<dyn Plugin>`],
+/// so methods on [`Plugin`] must not return function pointers.
+/// This includes `dyn Trait`, `fn(...)` syntax, `impl Trait`, and any types containing these.
+pub struct DynamicPlugin(Box<dyn Plugin>, Library);
```

Comment:
I definitely think we should use a named struct rather than a tuple struct here.

Reply:
agreed.